mobile/library/java/io/envoyproxy/envoymobile/engine/JvmCallbackContext.java

```diff
@@ -37,10 +37,7 @@ void passHeader(byte[] key, byte[] value, boolean start) {
   public Object onResponseHeaders(long headerCount, boolean endStream, long[] streamIntel) {
     assert bridgeUtility.validateCount(headerCount);
     final Map<String, List<String>> headers = bridgeUtility.retrieveHeaders();
-
-    callbacks.getExecutor().execute(
-        () -> callbacks.onHeaders(headers, endStream, new EnvoyStreamIntelImpl(streamIntel)));
-
```

Comment:
It's whichever thread that calls the callback, which is probably the worker thread? The idea for the original design is to not block the Envoy main thread, which makes sense. However, this design imposes the requirement that every callback must be executed in a separate thread from within the Envoy Mobile layer, which may not always be desirable depending on the needs. This change moves the responsibility to the `EnvoyHTTPCallbacks` implementer. For example, with Cronvoy as an implementer of the `EnvoyHTTPCallbacks`, it can decide when to execute the callback in a separate thread. If you noticed, in the Cronvoy implementation, it had `DirectExecutor` (it is removed in this PR), which is basically a no-op `Executor` because Cronvoy does not want every callback to be executed in a separate thread.

Reply:
So the Runnable task created in CronvoyUrlRequest.onData (https://github.com/envoyproxy/envoy/pull/32776/files#diff-8f04efffbc03f667d43c68783a3480f3b0b31cc61e136a07c778e5c177aca969R876) gets executed in the thread supplied by the Executor?

Reply:
Before this PR, there were two `Executor`s used in `CronvoyUrlRequest`:
- [DirectExecutor](https://github.com/envoyproxy/envoy/blob/da1a7d2ed07c6c60e7e0b05430337465ed91a03e/mobile/library/java/org/chromium/net/impl/CronvoyUrlRequest.java#L781-L784)
- [mUserExecutor](https://github.com/envoyproxy/envoy/blob/da1a7d2ed07c6c60e7e0b05430337465ed91a03e/mobile/library/java/org/chromium/net/impl/CronvoyUrlRequest.java#L153C26-L153C39)

The  method below came from `EnvoyHTTPCallbacks` and the [DirectExecutor](https://github.com/envoyproxy/envoy/blob/main/mobile/library/java/org/chromium/net/impl/CronvoyBidirectionalStream.java#L1112-L1117) implementation basically does nothing.

```
@Override
public Executor getExecutor() {
  return DIRECT_EXECUTOR;
}
```

The code in `CronvoyUrlRequest.onData` uses `execute`, which runs on [mUserExecutor](https://github.com/envoyproxy/envoy/blob/da1a7d2ed07c6c60e7e0b05430337465ed91a03e/mobile/library/java/org/chromium/net/impl/CronvoyUrlRequest.java#L616).

This PR removes `EnvoyHTTPCallbacks::getExecutor`. It won't matter for Cronvoy because returns a no-op `Executor` anyway for `EnvoyHTTPCallbacks::getExecutor`, but Cronvoy still uses `mUserExecutor` which is a user-supplied `Executor`. By not having `EnvoyHTTPCallbacks::getExecutor()`, Envoy Mobile no longer needs to perform a copy of the `ByteBuffer` before passing it to a user-defined `onResponseData` callback in the `JvmCallbackContext`.

From Envoy Mobile standpoint, we now only have `Envoy::Buffer::Instance` --> `envoy_data` and the data passed into `onResponseData` callback will be backed by `envoy_data` directly without any copy. The implementer of `EnvoyHTTPCallbacks::onData` can still make a copy if needed (just like the case with Cronvoy).

Reply:
so if we want to remove the enforcement that all up calls be run on the non-network thread, I'd think we'd want to ensure all current calls (unless there's exceptions) be run on the non-network thread.  I don't see any changes to the non-cronvoy up calls. Are we running those all on other threads already, or would this be a major regression for all non-cronvoy E-M?

Largely I think having an invariant to not run upcalls on the network thread is a good call.  Is there any other way to avoid the copy?
/wait-any

Reply:
> I don't see any changes to the non-cronvoy up calls.

Yeah, for Cronvoy, it uses its own `Executor` to run on the non-network thread: https://github.com/envoyproxy/envoy/blob/bd0130e7867937dec472afff0ad4a7b40a6f04b7/mobile/library/java/org/chromium/net/impl/CronvoyUrlRequest.java#L883-L901

> I don't see any changes to the non-cronvoy up calls. Are we running those all on other threads already, or would this be a major regression for all non-cronvoy E-M?

The idea is to let the users control that behavior whether or not they want to run the callbacks on the non-network thread. IOW, we don't want the library (the Java/Kotlin library in this case) to enforce that. If we look at the current implementation of [onResponseData](https://github.com/envoyproxy/envoy/blob/bd0130e7867937dec472afff0ad4a7b40a6f04b7/mobile/library/java/io/envoyproxy/envoymobile/engine/JvmCallbackContext.java#L73-L78), it currently does a copy of the `ByteBuffer` before passing it to the `Executor` even though the `Executor` implementation maybe a direct executor (no-op), which is the case the case with the [Cronvoy implementation](https://github.com/envoyproxy/envoy/blob/bd0130e7867937dec472afff0ad4a7b40a6f04b7/mobile/library/java/org/chromium/net/impl/CronvoyUrlRequest.java#L781-L784). From the Java/Kotlin library standpoint, there's no way to know if it uses a no-op `Executor`, so the safest way is to always create a copy.

Before this PR, Envoy Mobile would create 2 copies:
1. [JvmCallbackContext::onResponseData](https://github.com/envoyproxy/envoy/blob/bd0130e7867937dec472afff0ad4a7b40a6f04b7/mobile/library/java/io/envoyproxy/envoymobile/engine/JvmCallbackContext.java#L72-L79)
2. [CronvoyUrlRequest::onData](https://github.com/envoyproxy/envoy/blob/bd0130e7867937dec472afff0ad4a7b40a6f04b7/mobile/library/java/org/chromium/net/impl/CronvoyUrlRequest.java#L891)

After this PR, Cronvoy will only do one copy.

I agree that this change may inconvenience the users a little bit since now they have to decide whether or not to run certain callbacks on non-network threads, but at the same time the library will no longer need to always pay the penalty of creating a copy.

Another option that I can think of is changing the `EnvoyHTTPCallbacks::getExecutor` to something like
```
Optional<Executor> getExecutor
```
With this change, if the `Executor` is empty, we don't need to create a copy. For Cronvoy, we will set it to empty. Let me know what you think.

Reply:
"The idea is to let the users control that behavior whether or not they want to run the callbacks on the non-network thread." Hm, I'm not sure about this. While I can imagine that for some trivial applications this might be fine, I'm very concerned that this would be a massive footgun. I think Cronet runs callbacks on a different thread, IIRC, so it would seem like we should somehow be able to get the performance we need while preserving this invariant.

Reply:
Changing it to `Optional<Executor> getExecutor` is probably the best of both worlds? We can still default it to `Optional.of(Executors.newSingleThreadExecutor())`, but for Cronvoy, we will set it to `Optional.empty()`. When the `Executor` is empty, we will not need to do a copy (or the copy can be deferred to the implementer), but when the `Executor` is not empty, we will do a copy. What do you think?

Reply:
I'm usually weary of general APIs that take optional values, because most users of an API will truly treat them as optional and not look as closely at them, but this is a major decision that is being left on the user, so I think it's better not to make it optional.

Why does Cronvoy need to create it's own thread in CronvoyUrlRequest if it's already passing in an Executor to have the onResponseData function run in that thread?

To me, from the user's point of view, the ideal API would be:

onResponseData(lambda_to_run, executor_to_run_on)

That way, the invoker doesn't need to create it's own threads to run on, it just passes the executor and the lambda, and the implementation handles running the lambda on the executor with the response data.

I might be misunderstanding something.

As a side note: I agree the copying sucks so it would be nice to get rid of it, but I think the main repercussion of the copies would be memory pressure on the app.  I think the effect on CPU usage and latency would be minimal.

Reply:
> Why does Cronvoy need to create it's own thread in CronvoyUrlRequest if it's already passing in an Executor to have the onResponseData function run in that thread?

For the Cronvoy API, it wants to have control of the `Executor` [per-request](https://github.com/envoyproxy/envoy/blob/bd0130e7867937dec472afff0ad4a7b40a6f04b7/mobile/library/java/org/chromium/net/impl/CronvoyEngineBase.java#L94). This is not something that the current API supports. That's why Cronvoy implements a no-op `Executor` in the `EnvoyHTTPCallbacks` API.

> To me, from the user's point of view, the ideal API would be:
onResponseData(lambda_to_run, executor_to_run_on)

This can simply be implemented with
```
client.newStreamPrototype()
  .setOnResponseHeader { .... }
  .start(Optional.of(createFancyExecutor()))
```
 
By introducing an `Optional<Executor>`, the library **does not** always need to create a copy because it can know whether it needs to create a copy or not.

FTR, the default `Executor` in the existing code is a primitive [Executor.newSingleThreadExecutor](https://github.com/envoyproxy/envoy/blob/bd0130e7867937dec472afff0ad4a7b40a6f04b7/mobile/library/kotlin/io/envoyproxy/envoymobile/StreamPrototype.kt#L27), which is good for a toy project, but might  not be ideal when running on a real workload.